private_gpt/server/ingest/ingest_router.py

```diff
@@ -57,7 +73,11 @@ def ingest_file(request: Request, file: UploadFile) -> IngestResponse:
     service = request.state.injector.get(IngestService)
     if file.filename is None:
         raise HTTPException(400, "No file name provided")
-    ingested_documents = service.ingest_bin_data(file.filename, file.file)
+
+    metadata_dict = None if metadata is None else eval(metadata)
```

Comment:
I do not like how you're processing `metadata` field. Can you try to do a model or use a `dict`?
An example of function signature:
`def ingest_file(request: Request, file: UploadFile = File(...), metadata: Dict[str, str] = Body(...)) -> IngestResponse:`

Reply:
I tried to do it the way you suggested, but because the request contains a file upload, the request is handled differently and not processed as json, but as multipart/form-data, which prevents me from using a dict file type, I need to use a form field.

Reply:
What do you think about this proposal? It's a mock test, but it works. The only thing is that it takes `file` property as well.
```
async def extract_form_data(request: Request) -> Dict[str, Any]:
    form = await request.form()
    form_data = {key: value for key, value in form.items()}
    return form_data

@ingest_router.post("/ingest/file", tags=["Ingestion"])
def ingest_file(
    request: Request, file: UploadFile, metadata: dict[str, str] = Depends(extract_form_data)
) -> IngestResponse:
    ...
```

Reply:
This is a great suggestion, here is what i tried: 

```

async def extract_form_data(request: Request) -> Dict[str, Any] | None:
    form = await request.form()
    data = {key: value for key, value in form.items() if key != 'file'}
    return data["metadata"] if "metadata" in data else None

@ingest_router.post("/ingest/file", tags=["Ingestion"])
def ingest_file(
    request: Request, file: UploadFile, metadata: dict[str, Any] = Depends(extract_form_data)
) -> IngestResponse:
    
    service = request.state.injector.get(IngestService)
    
    if file.filename is None:
        raise HTTPException(400, "No file name provided") 
    
    print("metadata : " + str(metadata) + " type : " + str(type(metadata)))
    ingested_documents = service.ingest_bin_data(
        file.filename, file.file, metadata
    )
    return IngestResponse(object="list", model="private-gpt", data=ingested_documents)

```

The output from the print is metadata : {"foo":"bar"} type : <class 'str'>, so it is not intepreted as a dict. Digging deeper, it is not possible to have json encoded types such as a dict in a route that contains a File upload field ( see https://stackoverflow.com/questions/63110848/how-do-i-send-list-of-dictionary-as-body-parameter-together-with-files-in-fastap/71439821#71439821)

The only solutions I can think of is either to parse the string we receive as JSON, or to use eval().  
What's your opinion on these ?

Reply:
You can try to parse the string, it's better idea than `eval()` :)